pkg/storage/common.go

```diff
@@ -0,0 +1,413 @@
+		v, ok := manifest.Annotations[ispec.AnnotationRefName]
+		if ok && v == reference {
+			return manifest, true
+		}
+	}
+
+	return manifestDesc, false
+}
+
+func ValidateManifest(imgStore ImageStore, repo, reference, mediaType string, body []byte,
+	log zerolog.Logger,
+) (string, error) {
+	// validate the manifest
+	if !IsSupportedMediaType(mediaType) {
+		log.Debug().Interface("actual", mediaType).
+			Msg("bad manifest media type")
+
+		return "", zerr.ErrBadManifest
+	}
+
+	if len(body) == 0 {
+		log.Debug().Int("len", len(body)).Msg("invalid body length")
+
+		return "", zerr.ErrBadManifest
+	}
+
+	if mediaType == ispec.MediaTypeImageManifest {
+		var manifest ispec.Manifest
+		if err := json.Unmarshal(body, &manifest); err != nil {
+			log.Error().Err(err).Msg("unable to unmarshal JSON")
+
+			return "", zerr.ErrBadManifest
+		}
+
+		if manifest.Config.MediaType == ispec.MediaTypeImageConfig {
+			digest, err := validateOCIManifest(imgStore, repo, reference, &manifest, log)
+			if err != nil {
+				log.Error().Err(err).Msg("invalid oci image manifest")
+
+				return digest, err
+			}
+		}
+	} else if mediaType == artifactspec.MediaTypeArtifactManifest {
+		var m notation.Descriptor
+		if err := json.Unmarshal(body, &m); err != nil {
+			log.Error().Err(err).Msg("unable to unmarshal JSON")
+
+			return "", zerr.ErrBadManifest
+		}
+	}
+
+	return "", nil
+}
+
+func validateOCIManifest(imgStore ImageStore, repo, reference string, manifest *ispec.Manifest,
+	log zerolog.Logger,
+) (string, error) {
+	if manifest.SchemaVersion != SchemaVersion {
+		log.Error().Int("SchemaVersion", manifest.SchemaVersion).Msg("invalid manifest")
+
+		return "", zerr.ErrBadManifest
+	}
+
+	// validate image config
+	config := manifest.Config
+
+	blobFile, _, err := imgStore.GetBlob(repo, config.Digest.String(), "")
+	if err != nil {
+		return config.Digest.String(), zerr.ErrBlobNotFound
+	}
+
+	defer blobFile.Close()
+
+	dec := json.NewDecoder(blobFile)
+
+	var cspec ispec.Image
+	if err := dec.Decode(&cspec); err != nil {
+		return "", zerr.ErrBadManifest
+	}
+
+	// validate the layers
+	for _, l := range manifest.Layers {
+		blobFile, _, err := imgStore.GetBlob(repo, l.Digest.String(), "")
+		if err != nil {
+			return l.Digest.String(), zerr.ErrBlobNotFound
+		}
+
+		defer blobFile.Close()
+	}
+
+	return "", nil
+}
+
+func GetAndValidateRequestDigest(body []byte, digest string, log zerolog.Logger) (godigest.Digest, error) {
+	bodyDigest := godigest.FromBytes(body)
+
+	d, err := godigest.Parse(digest)
+	if err == nil {
+		if d.String() != bodyDigest.String() {
+			log.Error().Str("actual", bodyDigest.String()).Str("expected", d.String()).
+				Msg("manifest digest is not valid")
+
+			return "", zerr.ErrBadManifest
+		}
+	}
+
+	return bodyDigest, err
+}
+
+/* CheckIfIndexNeedsUpdate verifies if an index needs to be updated given a new manifest descriptor.
+Returns whether or not index needs update, in the later case it will also return the previous digest. */
```

Comment:
typo, maybe use "in the latter case it will also return"

Reply:
done, thanks